tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -1354,11 +1343,103 @@ LogicalResult emitFirstRoundShuffle(
   return success();
 }
 
+LogicalResult emitSecondRoundShuffleStitch(
+    OpBuilder& b, Location loc, Operation* op,
+    mlir::ResultRange::iterator acc_iter, Value thread_id_is_zero,
+    SmallVector<Value> output_idx_vec, int row_tile, int block_size,
```

Comment:
The difference between emitSecondRoundShuffleStitch and emitSecondRoundShuffle:
1, emitSecondRoundShuffle is able to accept a vector of row_reduction ops;
2, emitSecondRoundShuffleStitch is able to write into shared buffers.
emitSecondRoundShuffle can be seen as a special version of emitSecondRoundShuffleStitch if externel_output_only is set. Is it possible to merge the two functions?

Reply:
I create a new function is to support new features without break existing logics. The difference in the newest commit is more than what you noticed in the last commit after I add warp-wise reduction support.
My plan is to merge them in the next step, not current pr, after kStitch for fusion is robust enough to enable by default. And I also plan to combine block-wise and warp-wise schedules into one in that step.